tls/extensions/s2n_client_psk.c

```diff
@@ -162,16 +140,27 @@ static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn,
     ENSURE_REF(conn);
     ENSURE_REF(wire_identities_in);
 
+    uint16_t identities_count = 0;
+    GUARD_RESULT(s2n_count_psk_identities(wire_identities_in, &identities_count));
+
+    if (identities_count == 0) { 
+        return S2N_RESULT_OK; 
+    }
+
+    DEFER_CLEANUP(struct s2n_blob wire_identities_blob = { 0 }, s2n_free);
+    GUARD_AS_RESULT(s2n_alloc(&wire_identities_blob, identities_count * sizeof(struct s2n_psk_identity)));
```

Comment:
We could, but it would not be a limit based on the RFC, so theoretically we could receive a valid ClientHello with more identities than our limit allowed. These are the wire identities; the only limit the RFC places on them is that the length of the list has to fit in uint16_t.

In practice though, it seems unlikely that a client would need to send hundreds of identities, let alone thousands, so maybe it would be reasonable to limit what we accept over the wire.

Reply:
Doing some back-of-the-napkin math this means the peer can require us to allocate ~0.5MB in addition to the buffer required to store the raw identities. Another potential issue with this is we'll end up doing `N * M` comparisons where `N = peer supplied identities` and `M = length of the valid identities`.

Reply:
Yeah that seems like a good argument for limiting it, especially this early in the handshake. We could make the limit something low like 100 by default, and potentially make it configurable later.

This also seems like another reason to special-case our default behavior like I suggest in https://github.com/awslabs/s2n/pull/2512#discussion_r562244471. We should be iterating over our local list of PSKs, not the wire list.

Reply:
Synced up offline: We decided to add an upper limit of 100 psk identities and perform a <= safety check  prior to allocating memory on line 151.